api/client/src/main/java/org/projectnessie/client/auth/oauth2/AuthorizationCodeFlow.java

```diff
@@ -0,0 +1,249 @@
+import java.io.PrintStream;
+import java.io.UncheckedIOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.projectnessie.client.http.HttpClient;
+import org.projectnessie.client.http.HttpRequest;
+import org.projectnessie.client.http.HttpResponse;
+import org.projectnessie.client.http.impl.HttpUtils;
+import org.projectnessie.client.http.impl.UriBuilder;
+
+class AuthorizationCodeFlow implements AutoCloseable {
+
+  static final String CONTEXT_PATH = "/nessie-client/auth";
+  static final String MSG_PREFIX = "[nessie-oauth2-client] ";
+
+  private static final org.slf4j.Logger LOGGER =
+      org.slf4j.LoggerFactory.getLogger(AuthorizationCodeFlow.class);
+
+  private static final String HTML_TEMPLATE_OK =
+      "<html><body><h1>Authentication successful</h1><p>You can close this page now.</p></body></html>";
+  private static final String HTML_TEMPLATE_FAILED =
+      "<html><body><h1>Authentication failed</h1><p>Could not obtain access token: %s</p></body></html>";
+
+  private static final int STATE_LENGTH = 16;
+
+  private final OAuth2ClientParams params;
+  private final HttpClient httpClient;
+  private final PrintStream console;
+  private final String state;
+  private final HttpServer server;
+  private final String redirectUri;
+  private final URI authorizationUri;
+  private final Duration flowTimeout;
+
+  private final CompletableFuture<HttpExchange> requestFuture = new CompletableFuture<>();
+  private final CompletableFuture<Void> responseFuture = new CompletableFuture<>();
+  private final CompletableFuture<Tokens> tokensFuture;
+  private final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+
+  AuthorizationCodeFlow(OAuth2ClientParams params, HttpClient httpClient) {
+    this(params, httpClient, System.out);
+  }
+
+  /** Constructor used for testing. */
+  AuthorizationCodeFlow(OAuth2ClientParams params, PrintStream console) {
+    this(params, params.getHttpClient().build(), console);
+  }
+
+  private AuthorizationCodeFlow(
+      OAuth2ClientParams params, HttpClient httpClient, PrintStream console) {
+    this.params = params;
+    this.httpClient = httpClient;
+    this.console = console;
+    this.flowTimeout = params.getAuthorizationCodeFlowTimeout();
+    // Close the server when the close future completes, but wait for the response to be sent first,
+    // so that the browser does not show "Connection closed" to the user.
+    CompletableFuture.allOf(closeFuture, responseFuture).thenRun(this::doClose);
+    tokensFuture =
+        requestFuture.thenApply(this::extractAuthorizationCode).thenApply(this::fetchNewTokens);
```

Comment:
Not quite, `requestFuture` completes when an incoming request arrives, but _before_ we extract the auth code. I introduced this future to prevent the web server from treating subsequent requests (e.g. if the browser refreshes): once we have a code, we use it, and it won't change, even if a different code arrives later on (unlikely, but who knows).

I actually haven't materialized the future that completes when the code is extracted, but indeed I can do that.

Reply:
no worries. I think the code if fine "as is".

Reply:
Too late :-) It's probably a good thing to materialize that future anyway: if we plan to introduce other ways to grab the code (e.g. from `System.in`), then that future should be completed from many places, and the first one would win.